### PR TITLE
Adding support for properties encoded in VNN-LIB format into the core C++ side of Marabou

### PR DESCRIPTION
**Pull request by idan0610**

### CODE REVIEW

**Comment by wu-haoze** [comment]
Hi @idan0610 , I recall there is a vnn lib parser implemented for Python. Does this PR replace that one?

**Comment by idan0610** [comment]
Hi @wu-haoze , for now this PR is in addition to the PR that added the Python support, as there was a requirement to add the C++ support as well (when the work on the Python side was nearing completion). Of course, it is possible in the future to remove the Python implementation, and to use this implementation on the Python side as well.

File: src/common/MString.cpp [hunk 1]
```diff
@@ -155,13 +155,13 @@ String String::trim() const
 
     for ( unsigned i = 0; i < length(); ++i )
     {
-        if ( ( !firstNonSpaceFound ) && ( _super[i] != ' ' ) && ( _super[i] != '\n' ) )
+        if ( ( !firstNonSpaceFound ) && ( _super[i] != ' ' ) && ( _super[i] != '\n' ) && ( _super[i] != '\r' ) )
```
Comment (guykatzz) [inline src/common/MString.cpp:158]: Please adjust MString's test to test for this new kind of trimming as well

File: src/engine/DnCMarabou.cpp [hunk 2]
```diff
@@ -20,6 +20,7 @@
 #include "MStringf.h"
 #include "Options.h"
 #include "PropertyParser.h"
+#include "VnnLibParser.h"
```
Comment (guykatzz) [inline src/engine/DnCMarabou.cpp:23]: The "includes" should be sorted alphabetically

File: src/input_parsers/VnnLibParser.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+ ** directory for licensing information.\endverbatim
+ **
+ ** A parser for properties encoded in VNN-LIB format.
+
+**/
+
+#ifndef _VnnLibParser_h_
+#define _VnnLibParser_h_
+
+#include "InputQuery.h"
```
Comment (guykatzz) [inline src/input_parsers/VnnLibParser.h:19]: Sort alphabetically

File: src/input_parsers/VnnLibParser.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+#include "MString.h"
+#include "Vector.h"
+#include "Map.h"
+
+class VnnLibParser
+{
+    public:
```
Comment (guykatzz) [inline src/input_parsers/VnnLibParser.h:26]: The indentation is off. "public:" and "private:" statements should not be indented. Please compare to other files

File: src/input_parsers/tests/Test_VnnLibParser.h [hunk 5]
```diff
@@ -0,0 +1,719 @@
...
+ **/
+
+#include <cxxtest/TestSuite.h>
+
+#include "Engine.h"
+#include "InputQuery.h"
+#include "OnnxParser.h"
+#include "VnnLibParser.h"
+#include <filesystem>
+
```
Comment (guykatzz) [inline src/input_parsers/tests/Test_VnnLibParser.h:22]: So, all of these are great tests, but they are more regression tests than unit tests.
(I see that the same issue happens with the onnx parser)

Please make this a regression test, and leave the unit test empty for now - and then we can add these tests together, using the CxxTest mock infrastructure.

File: src/input_parsers/VnnLibParser.cpp [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+
+**/
+
+#include <regex>
+#include "File.h"
+#include "InputParserError.h"
+#include "VnnLibParser.h"
+#include "DisjunctionConstraint.h"
+
+static double extractScalar ( const String &token )
```
Comment (guykatzz) [inline src/input_parsers/VnnLibParser.cpp:22]: no space between function name and (

File: src/input_parsers/VnnLibParser.cpp [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+{
+    std::string::size_type end;
+    double value = std::stod ( token.ascii (), &end );
```
Comment (guykatzz) [inline src/input_parsers/VnnLibParser.cpp:25]: again, no space before (

File: src/input_parsers/VnnLibParser.cpp [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+    if ( end != token.length ())
```
Comment (guykatzz) [inline src/input_parsers/VnnLibParser.cpp:26]: .length ()) --> .length() )

File: src/input_parsers/VnnLibParser.cpp [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+        // A "READ_FAILED" is how we know we're out of lines
+        if ( e.getCode () != CommonError::READ_FAILED )
+            throw e;
+    }
+
+    return vnnlibContent;
+}
+
+void
+VnnLibParser::parse ( const String &vnnlibFilePath, InputQuery &inputQuery )
```
Comment (guykatzz) [inline src/input_parsers/VnnLibParser.cpp:71]: return type should appear on the same line as the function

File: src/input_parsers/VnnLibParser.cpp [hunk 10]
```diff
@@ -0,0 +1,549 @@
...
+{
+    int index = 0;
+
+    while (( unsigned ) index < tokens.size ())
+    {
+        ASSERT( tokens[index] == "(" )
+        index = parseCommand ( index + 1, tokens, inputQuery );
+        ASSERT( tokens[index] == ")" )
+
+        index++;
```
Comment (guykatzz) [inline src/input_parsers/VnnLibParser.cpp:113]: always prefer ++index to index++ when both can be used; it is more efficient, generally

File: src/input_parsers/VnnLibParser.cpp [hunk 11]
```diff
@@ -0,0 +1,536 @@
...
+    }
+
+    parseScript( all_tokens, inputQuery );
+}
+
+int VnnLibParser::parseScript( const Vector<String> &tokens, InputQuery &inputQuery )
+{
+    int index = 0;
+
+    while (( unsigned ) index < tokens.size() )
```
Comment (guykatzz) [inline src/input_parsers/VnnLibParser.cpp:104]: (( unsigned ) index -> ( (unsigned)index )

File: src/input_parsers/VnnLibParser.cpp [hunk 12]
```diff
@@ -0,0 +1,536 @@
...
+    {
+        throw InputParserError( InputParserError::UNEXPECTED_INPUT, command_name.ascii() );
+    }
+
+    return index;
+}
+
+int VnnLibParser::parseDeclareConst( int index, const Vector<String> &tokens, InputQuery &inputQuery )
+{
+    const String &varName = tokens[index++];
```
Comment (guykatzz) [inline src/input_parsers/VnnLibParser.cpp:138]: index++ --> ++index, on both lines

**Review by guykatzz — APPROVED** [review]
Really minor comments, and then you can merge. Thanks!

**Pull request merged**